### PR TITLE
feat:  Add support for git hash and sorting the environment list on App and App Group Overview page 

### PR DESCRIPTION
**Pull request by RajeevRanjan27**
# Description

Fixes #

## Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/app/AppListingService.go [hunk 1]
```diff
@@ -768,6 +797,84 @@ func (impl AppListingServiceImpl) FetchAppStageStatus(appId int, appType int) ([
...
+	// Retrieve workflows associated with the artifact IDs, handling any potential error.
+	artifactsWithGitTriggers, err := impl.ciWorkflowRepository.FindAllLastGitTriggeredWorkflowByArtifactIds(ciArtifactIds)
+	if err != nil {
+		// Log the error along with the artifact IDs that caused it.
+		impl.Logger.Errorw("error retrieving GitTriggers of the  CiWorkflows", "ciArtifactIds", ciArtifactIds, "err", err)
+		return nil, err // Return nil map and the encountered error.
+	}
+
+	// Map to hold the last Git-triggered workflow for each artifact.
+	gitTriggersToParentArtifactMap := make(map[int]map[int]pipelineConfig.GitCommit)
```
Comment (Ash-exp) [inline pkg/app/AppListingService.go:833]: define a **GitTriggers** type of `map[int]pipelineConfig.GitCommit` and use it

File: pkg/app/AppListingService.go [hunk 2]
```diff
@@ -768,6 +797,84 @@ func (impl AppListingServiceImpl) FetchAppStageStatus(appId int, appType int) ([
...
+
+		// Iterate through the Git triggers and extract the commit information.
+		for _, git := range ciWorkflowTemp {
+			gitCommits = append(gitCommits, git.Commit)
+		}
+
+		// Append the extracted commits to the corresponding child artifact in the map.
+		gitCommitsWithArtifactMap[child] = append(gitCommitsWithArtifactMap[child], gitCommits...)
+
+		// Ensure that gitCommitsWithArtifactMap[child] is initialized if it doesn't exist.
```
Comment (Ash-exp) [inline pkg/app/AppListingService.go:870]: this logic should be prior to 
```
gitCommitsWithArtifactMap[child] = append(gitCommitsWithArtifactMap[child], gitCommits...)
```

File: pkg/app/AppListingService.go [hunk 3]
```diff
@@ -792,6 +899,23 @@ func (impl AppListingServiceImpl) FetchOtherEnvironment(ctx context.Context, app
 		impl.Logger.Errorw("error in fetching latest chart", "err", err)
 		return envs, err
 	}
+	ciArtifactsWithParent := make([]*CiArtifactWithParentArtifact, 0)
+	for _, env := range envs {
+
+		if ok := env.CiArtifactId; ok > 0 {
```
Comment (Ash-exp) [inline pkg/app/AppListingService.go:905]: simplify to `env.CiArtifactId > 0`

File: pkg/app/AppListingService.go [hunk 4]
```diff
@@ -792,6 +899,23 @@ func (impl AppListingServiceImpl) FetchOtherEnvironment(ctx context.Context, app
...
+			ciArtifactsWithParent = append(ciArtifactsWithParent, &CiArtifactWithParentArtifact{
```
Comment (Ash-exp) [inline pkg/app/AppListingService.go:906]: use explicit assignments to `CiArtifactWithParentArtifact`

File: pkg/app/AppListingService.go [hunk 5]
```diff
@@ -792,6 +899,23 @@ func (impl AppListingServiceImpl) FetchOtherEnvironment(ctx context.Context, app
...
+				env.ParentCiArtifactId, env.CiArtifactId,
+			})
+		}
+
+	}
+
+	gitCommitsWithArtifacts, err := impl.FetchCiArtifactToGitTriggersMap(ciArtifactsWithParent)
+	if err != nil {
+		impl.Logger.Errorw("Error in fetching the git commits of the ciArtifacts")
```
Comment (Ash-exp) [inline pkg/app/AppListingService.go:915]: add meta data `ciArtifactsWithParent` to log

File: pkg/app/AppListingService.go [hunk 6]
```diff
@@ -805,6 +929,15 @@ func (impl AppListingServiceImpl) FetchOtherEnvironment(ctx context.Context, app
 		} else {
 			env.ChartRefId = chart.ChartRefId
 		}
+
+		gitCommits, exists := gitCommitsWithArtifacts[env.CiArtifactId]
```
Comment (Ash-exp) [inline pkg/app/AppListingService.go:933]: can we make util function to do the below things using generics

File: pkg/plugin/GlobalPluginService.go [hunk 7]
```diff
@@ -1345,6 +1345,34 @@ func filterPluginStepData(existingPluginStepsInDb []*repository.PluginStep, plug
 		return nil, nil, pluginStepUpdateReq
 	}
 
+	//newPluginStepsToCreate := make([]*PluginStepsDto, 0)
```
Comment (Ash-exp) [inline pkg/plugin/GlobalPluginService.go:1348]: remove commented code

**Review by Ash-exp — CHANGES_REQUESTED** [review]

File: internal/sql/repository/AppListingRepository.go [hunk 8]
```diff
@@ -88,6 +90,11 @@ type AppListingRepositoryImpl struct {
 	gitOpsRepository                 GitOpsConfigRepository
 }
 
+type CiWorkflowRepositoryImpl struct {
```
[the remark this answers is not in the document]
Comment (RajeevRanjan27) [reply internal/sql/repository/AppListingRepository.go:93]: resolved

File: internal/sql/repository/pipelineConfig/CiWorkflowRepository.go [hunk 9]
```diff
@@ -48,6 +48,8 @@ type CiWorkflowRepository interface {
 	ExistsByStatus(status string) (bool, error)
 	FindBuildTypeAndStatusDataOfLast1Day() []*BuildTypeCount
 	FIndCiWorkflowStatusesByAppId(appId int) ([]*CiWorkflowStatus, error)
+	FindAllLastGitTriggeredWorkflowByArtifactIds(ciArtifactIds []int) (ciWorkflows []*ArtifactAndGitCommitMapping, err error)
```
[the remark this answers is not in the document]
Comment (RajeevRanjan27) [reply internal/sql/repository/pipelineConfig/CiWorkflowRepository.go:51]: resolved

**Pull request merged**